### PR TITLE
perl5db.pl: "c subname" and "b subname" break on first executable line if possible

### LINKED ISSUE #799: Debugger:  Break gets set on wrong subroutine
Migrated from [rt.perl.org#1726](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=1726) (status was 'open')

Searchable as RT1726$

### PR DESCRIPTION
**Pull request by tonycoz**
Fixes #799

### CODE REVIEW

File: lib/perl5db.pl [hunk 1]
```diff
@@ -2033,19 +2033,12 @@ sub _DB__handle_c_command {
...
 
-            # find_sub will return "file:line_number" corresponding
-            # to where the subroutine is defined; we call find_sub,
-            # break up the return value, and assign it in one
-            # operation.
-            ( $file, $i ) = ( find_sub($subname) =~ /^(.*):(.*)$/ );
-
-            # Force the line number to be numeric.
-            $i = $i + 0;
+            my ($file, $s, $e) = do { subroutine_first_breakable_line($subname) };
```
Comment (mauke) [inline lib/perl5db.pl]: Did you mean for this `do` to be an `eval`?

File: lib/perl5db.pl [hunk 2]
```diff
@@ -2058,19 +2051,18 @@ sub _DB__handle_c_command {
...
                 }
                 $i = $_line_num;
             } ## end if ($i)
-
-            # We didn't find a sub by that name.
             else {
-                print $OUT "Subroutine $subname not found.\n";
-                next CMD;
+                  print $OUT $@;
+                  next CMD;
```
Comment (mauke) [inline lib/perl5db.pl]: Why the extra indentation here?

File: lib/perl5db.pl [hunk 3]
```diff
@@ -5389,6 +5381,61 @@ sub subroutine_filename_lines {
     return (find_sub($subname) =~ /^(.*):(\d+)-(\d+)$/);
 } ## end sub subroutine_filename_lines
 
+=head2 subroutine_first_breakable(subname)
```
Comment (mauke) [inline lib/perl5db.pl]: Did you mean: `subroutine_first_breakable_line`?

File: lib/perl5db.pl [hunk 4]
```diff
@@ -5389,6 +5381,61 @@ sub subroutine_filename_lines {
...
+
+Return the file and line number of the first breakable line in
+subname.
```
Comment (mauke) [inline lib/perl5db.pl]: It actually returns three values.

File: lib/perl5db.pl [hunk 5]
```diff
@@ -5401,9 +5448,8 @@ sub break_subroutine {
     my $subname = shift;
 
     # Get filename, start, and end.
-    my ( $file, $s, $e ) = subroutine_filename_lines($subname)
-      or die "Subroutine $subname not found.\n";
-
+    my ( $file, $s, $e ) = eval { subroutine_first_breakable_line($subname) }
+      or die $@;
```
Comment (mauke) [inline lib/perl5db.pl]: `eval { foo() } or die $@` could be simplified to just `foo()`.

**Comment by tonycoz** [comment]
What I pushed last week was a bit undercooked.

@mauke, thanks for the feedback, I think I've fixed each of the problems you mentioned, and a few more.

File: lib/perl5db.pl [hunk 6]
```diff
@@ -5389,6 +5381,60 @@ sub subroutine_filename_lines {
...
 } ## end sub subroutine_filename_lines
 
+=head2 subroutine_first_breakable_line(subname)
+
+Return the file and a range of breakable line numbers in subname, as a list:
+
+  my ($file, $sline, $eline) = subroutine_first_breakable_line($subname);
+
+It first tries to find the line number and file of the first OP in
+subname, if that fails it falls back to returning the range of lines
```
Comment (mauke) [inline lib/perl5db.pl]: I'd punctuate this as `... subname. If that fails, it falls back ...`.

I feel like the second sentence is missing something. Maybe "... the range of all lines covered by subname"? My understanding at least is that we first try to be precise and return the line of the first OP exactly; if that fails, we give up and just return the range of all line numbers that lie within the given sub (and our caller has to deal with it).

Comment (tonycoz) [reply lib/perl5db.pl]: Thinking about it, the way the function behaved made it difficult to document (and use), so it's behaviour is now closer to what the name promises, and now leaves less work for callers.

File: lib/perl5db.pl [hunk 7]
```diff
@@ -2055,22 +2048,13 @@ sub _DB__handle_c_command {
...
-                # after the 'sub whatever' line.
-                $max = $#dbline;
-                my $_line_num = $i;
-                while ($dbline[$_line_num] == 0 && $_line_num< $max)
-                {
-                    $_line_num++;
-                }
-                $i = $_line_num;
+                $i = $line;
             } ## end if ($i)
```
Comment (mauke) [inline lib/perl5db.pl]: Just noticed this: The comment should be either removed or changed to `end if ($line)`.

**Comment by mauke** [comment]
Looks good to me.

**Comment by tonycoz** [comment]
added b postpone support

**Review by mauke — APPROVED** [review]

**Pull request merged**